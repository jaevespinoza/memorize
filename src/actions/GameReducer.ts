// src/features/gameSlice.js
import { createSlice } from "@reduxjs/toolkit";
import { ISelectedImage } from "../interfaces/AppActionsInterface";

export interface IGameState {
  selectedCards: ISelectedImage[];
  foundCards: ISelectedImage[];
  matchPopup: boolean;
  matchedCard: boolean;
  successfulGames: number;
  success: number;
  errors: number;
  name: string;
}

const gameSlice = createSlice({
  name: "gameApplication",
  initialState: {
    selectedCards: [] as ISelectedImage[],
    foundCards: [] as ISelectedImage[],
    success: 0,
    matchPopup: false,
    matchedCard: false,
    errors: 0,
    name: "",
    successfulGames: 0,
  } as IGameState,
  reducers: {
    /**
     * Sets whether the cards match or not
     */
    setMatchedCard: (state, action) => {
      state.matchedCard = action.payload;
    },
    /**
     * Sets whether the match alert should be shown or not
     */
    setMatchedPopup: (state, action) => {
      state.matchPopup = action.payload;
    },
    /**
     * Sets the selected cards into the state and then checks
     * whether they match or not
     */
    setSelectedCards: (state, action) => {
      if (!state.selectedCards.some((item) => item.id === action.payload.id)) {
        state.selectedCards.push(action.payload);
        checkSelected(state);
      }
    },
    /**
     * Sets the found cards into the state so they are still shown
     */
    setFoundCards: (state, action) => {
      state.foundCards.push(action.payload);
    },
    /**
     * Sets the number of successful matches
     */
    setSuccess: (state, action) => {
      state.success = action.payload;
    },
    /**
     * Sets the number of wrong matches
     */
    setErrors: (state, action) => {
      state.errors = action.payload;
    },
    /**
     * Sets the name of the user into the state
     */
    setName: (state, action) => {
      state.name = action.payload;
    },
    /**
     * Resets the selected cards and then adds to the error
     */
    resetCards: (state) => {
      state.selectedCards = [];
    },
    resetGame: (state) => {
      state.foundCards = [];
      state.selectedCards = [];
      state.success = 0;
      state.errors = 0;
      state.successfulGames += 1;
    },
  },
});

/**
 * Checks the cards that were selected.
 * If the cards match, they are added to the pool of found cards and
 * the success popup shows. If not, they are discarded and the
 * error popup shows.
 * @param state The Application state
 * @returns The updated state with the amount of errors and successes updated
 */
const checkSelected = (state: IGameState) => {
  // If the user hasn't selected more than two cards, we return
  if (state.selectedCards.length < 2) {
    return;
  }
  // We compare the names of the cards
  const checkSelectedCards =
    state.selectedCards[0].name === state.selectedCards[1].name;
  state.matchPopup = true;
  // If they match, we set the matchedCard variable to true. If not, it's false.
  if (checkSelectedCards) {
    state.matchedCard = true;
    successOnMatch(state);
  } else {
    state.matchedCard = false;
    state.errors += 1;
  }
};

const successOnMatch = (state: IGameState) => {
  state.foundCards.push(state.selectedCards[0]);
  state.foundCards.push(state.selectedCards[1]);
  state.selectedCards = [];
  state.success += 1;
};

export const {
  setSelectedCards,
  setFoundCards,
  setMatchedCard,
  resetCards,
  resetGame,
  setMatchedPopup,
  setSuccess,
  setErrors,
  setName,
} = gameSlice.actions;
export default gameSlice.reducer;
