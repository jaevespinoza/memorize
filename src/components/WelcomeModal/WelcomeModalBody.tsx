interface IWelcomeModalBody {
  onSaveName: () => void;
  nameInput: string;
  setNameInput: (name: string) => void;
}

/**
 * Body of the welcome modal that will hold the form to
 * write the name of the user
 * @param onSaveName Function to execute after the name is submitted
 * @param nameInput Input of the name written by the user
 * @param setNameInput Function that saves the input into the component's state
 */
const WelcomeModalBody = ({
  onSaveName,
  nameInput,
  setNameInput,
}: IWelcomeModalBody) => {
  /**
   * Upon pressing the "Enter" key, the name should be saved.
   * It also makes sure that the name input exists
   * @param event Key press event
   */
  const enterKeyHandler = (event: React.KeyboardEvent) => {
    if (event.key === "Enter" && nameInput.length > 0) {
      onSaveName();
      event.preventDefault();
    }
  };

  return (
    <>
      <p>
        To ensure a better experience, please write your name in the field
        below:
      </p>
      <form>
        <div className="mb-3">
          <label htmlFor="fullName" className="form-label">
            Full name
          </label>
          <input
            type="text"
            onChange={(event) => setNameInput(event.currentTarget.value)}
            onKeyDown={enterKeyHandler}
            className="form-control"
            id="fullName"
            aria-describedby="emailHelp"
          />
        </div>
      </form>
    </>
  );
};

export default WelcomeModalBody;
