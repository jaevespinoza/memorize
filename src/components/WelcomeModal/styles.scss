.modal {
  transition: opacity 0.3s ease-in-out;
  /* Additional modal styles go here */
}

.modal.fade.show {
  opacity: 1;
}

.welcome-modal {
  &__content {
    height: 350px;
  }
  &__header {
  }
  &__body {
    display: flex;
    flex-direction: column;
    justify-content: center;
  }
  &__footer {
    height: 75px;
    &__button {
      &:disabled {
        background-color: rgba(0, 0, 0, 0.09);
        border: none;
        color: #000;
      }
    }
  }
}
