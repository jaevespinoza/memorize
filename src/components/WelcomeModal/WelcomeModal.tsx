import { useEffect, useState } from "react";
import { Modal, Button, Form, Col, Container } from "react-bootstrap";
import { setName } from "../../actions/GameReducer";
import "./styles.scss";

const WelcomeModal = () => {
  const [nameInput, setNameInput] = useState("");
  const [show, setShow] = useState(false);

  useEffect(() => {
    setShow(true);
  }, []);

  const enterKeyHandler = (event: React.KeyboardEvent) => {
    if (event.key === "Enter" && nameInput.length > 0) {
      setName(nameInput);
      event.preventDefault();
    }
  };

  const onSaveName = () => {};

  return (
    <>
      <div
        className={`modal fade ${show ? "show" : ""}`}
        style={{ display: show ? "block" : "none" }}
        tabIndex={-1}
        role="dialog"
        data-bs-backdrop="static"
        data-bs-keyboard="false"
      >
        <div className="modal-dialog ">
          <div className="modal-content welcome-modal__content">
            <div className="modal-header">
              <h5 className="modal-title">Welcome to Memorize!</h5>
            </div>
            <div className="modal-body welcome-modal__body">
              <p>
                To ensure a better experience, please write your name in the
                field below:
              </p>
              <form>
                <div className="mb-3">
                  <label htmlFor="fullName" className="form-label">
                    Full name
                  </label>
                  <input
                    type="text"
                    onChange={(event) =>
                      setNameInput(event.currentTarget.value)
                    }
                    className="form-control"
                    id="fullName"
                    aria-describedby="emailHelp"
                  />
                </div>
              </form>
            </div>
            <div className="modal-footer welcome-modal__footer">
              <button
                type="button"
                className="btn btn-primary welcome-modal__footer__button"
                disabled={nameInput.length === 0}
                onClick={onSaveName}
              >
                Save changes
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className={`modal-backdrop fade ${show ? "show" : ""}`}
        style={{ display: show ? "block" : "none" }}
      ></div>
    </>
  );

  /*
  return (
    <Modal
      show={show}
      onHide={() => setShow(false)}
      className={styles.welcome_modal}
      backdrop="static"
      size="lg"
      aria-labelledby="contained-modal-title-vcenter"
      centered
    >
      <Modal.Header>
        <Modal.Title>Welcome to Concentration!</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Container>
          <Col>
            To fully enjoy this game, please write your full name on the form
            below:
          </Col>
          <Col>
            <Form.Control
              size="sm"
              value={nameInput}
              onChange={(event) => {
                setNameInput(event.currentTarget.value);
              }}
              onKeyDown={enterKeyHandler}
              type="text"
              placeholder="Write your name here"
            />
          </Col>
        </Container>
      </Modal.Body>
      <Modal.Footer>
        <Button
          variant="primary"
          onClick={() => setName(nameInput)}
          disabled={nameInput.length === 0}
        >
          Save name
        </Button>
      </Modal.Footer>
    </Modal>
  ); */
};

export default WelcomeModal;
