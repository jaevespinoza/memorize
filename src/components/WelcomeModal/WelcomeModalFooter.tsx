import { useTranslation } from "react-i18next";
import "./styles.scss";

interface IWelcomeModalFooter {
  onSaveName: () => void;
  enabledButton: boolean;
}

/**
 * Content of the footer that will be in the welcome dialog
 * @param onSaveName Function to be executed when clicking the button
 * @param enabledButton Boolean that determines whether the button is enabled or not
 */
const WelcomeModalFooter = ({
  onSaveName,
  enabledButton,
}: IWelcomeModalFooter) => {
  const { t } = useTranslation("game");
  return (
    <button
      type="button"
      className="btn btn-primary welcome-modal__footer__button"
      disabled={!enabledButton}
      onClick={onSaveName}
      data-testid="save-button"
    >
      {t("welcome.save")}
    </button>
  );
};

export default WelcomeModalFooter;
