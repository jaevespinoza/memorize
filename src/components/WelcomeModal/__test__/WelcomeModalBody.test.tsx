import { configureStore } from "@reduxjs/toolkit";
import { ToolkitStore } from "@reduxjs/toolkit/dist/configureStore";
import { beforeEach, describe, expect, test } from "vitest";
import GameReducer, { IGameState } from "../../../actions/GameReducer";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import { render, screen } from "@testing-library/react";
import { I18nextProvider } from "react-i18next";
import i18n from "../../../i18n/i18n";
import WelcomeModal from "../WelcomeModal";
import userEvent from "@testing-library/user-event";

describe("Welcome Modal", () => {
  let store: ToolkitStore<{ game: IGameState }>;

  beforeEach(() => {
    store = configureStore({ reducer: { game: GameReducer } });
  });

  test("matches snapshot", () => {
    const body = render(
      <Provider store={store}>
        <I18nextProvider i18n={i18n}>
          <WelcomeModal />
        </I18nextProvider>
      </Provider>
    );

    expect(body).toMatchSnapshot();
  });

  test("empty input", () => {
    render(
      <Provider store={store}>
        <I18nextProvider i18n={i18n}>
          <WelcomeModal />
        </I18nextProvider>
      </Provider>
    );

    expect(screen.getByRole("form")).toHaveFormValues({
      fullName: "",
    });
  });

  test("changes the input", () => {
    render(
      <Provider store={store}>
        <I18nextProvider i18n={i18n}>
          <WelcomeModal />
        </I18nextProvider>
      </Provider>
    );

    userEvent
      .type(screen.getByTestId("name-input"), "Javier")
      .then(() =>
        expect(screen.getByTestId("name-input")).toHaveValue("Javier")
      );
  });
});
