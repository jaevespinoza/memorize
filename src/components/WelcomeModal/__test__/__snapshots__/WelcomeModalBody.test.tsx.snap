// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Welcome Modal > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="modal fade show"
        data-bs-backdrop="static"
        data-bs-keyboard="false"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog "
        >
          <div
            class="modal-content welcome-modal__content"
          >
            <div
              class="modal-header"
            >
              <h5
                class="modal-title"
              >
                Welcome to Memorize!
              </h5>
            </div>
            <div
              class="modal-body welcome-modal__body"
            >
              <p>
                To ensure a better experience, please write your name in the field below:
              </p>
              <form
                aria-label="form"
                data-testid="form-input"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="fullName"
                  >
                    Full name
                  </label>
                  <input
                    aria-describedby="emailHelp"
                    aria-label="data-input"
                    class="form-control"
                    data-testid="name-input"
                    id="fullName"
                    name="fullName"
                    type="text"
                  />
                </div>
              </form>
            </div>
            <div
              class="modal-footer welcome-modal__footer"
            >
              <button
                class="btn btn-primary welcome-modal__footer__button"
                data-testid="save-button"
                disabled=""
                type="button"
              >
                Save name
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="modal-backdrop fade show"
        style="display: block;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="modal fade show"
      data-bs-backdrop="static"
      data-bs-keyboard="false"
      role="dialog"
      style="display: block;"
      tabindex="-1"
    >
      <div
        class="modal-dialog "
      >
        <div
          class="modal-content welcome-modal__content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Welcome to Memorize!
            </h5>
          </div>
          <div
            class="modal-body welcome-modal__body"
          >
            <p>
              To ensure a better experience, please write your name in the field below:
            </p>
            <form
              aria-label="form"
              data-testid="form-input"
            >
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="fullName"
                >
                  Full name
                </label>
                <input
                  aria-describedby="emailHelp"
                  aria-label="data-input"
                  class="form-control"
                  data-testid="name-input"
                  id="fullName"
                  name="fullName"
                  type="text"
                />
              </div>
            </form>
          </div>
          <div
            class="modal-footer welcome-modal__footer"
          >
            <button
              class="btn btn-primary welcome-modal__footer__button"
              data-testid="save-button"
              disabled=""
              type="button"
            >
              Save name
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="modal-backdrop fade show"
      style="display: block;"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
